### PR TITLE
201 vm linux jupyterhub - Azure4Students

### PR DESCRIPTION
**Pull request by leestott**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+                "publicIpAllocationMethod": "Dynamic"
+            },
+            "sku": {
+                "name": "Basic",
+                "tier": "Regional"
+            }
+        },
+        {
+            
+            "type": "Microsoft.Storage/storageAccounts",
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: how is this used?

Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: {
            
            "type": "Microsoft.Storage/storageAccounts",
            "apiVersion": "2019-06-01",
            "name": "[variables('storageAccountName')]",
            "location": "[parameters('location')]",
            "sku": {
                "name": "[variables('storageAccountType')]"
            },
            "kind": "[variables('storageAccountKind')]"
        },

Variables are

Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: Variables are 

   "storageAccountName": "[concat('storage', uniqueString(resourceGroup().id))]",
    "storageAccountType": "Standard_LRS",
    "storageAccountKind": "Storage",

Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: @bmoore-msft  see comments

Comment (bmoore-msft) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: sorry - why is there a storageAccount provisioned in this sample... it's not used anywhere.

Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: :-) yes I have added storage selection choices now and also optimised the VM Sizes based on feedback and requirements. Storage sizes are selected from a list now based on feedback from users.

Comment (bmoore-msft) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: Not sure I follow (you didn't answer the question)... why are you provisioning a storageAccount?

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+            "CPU-128GB": "Standard_E16s_v3",
+            "CPU-256Gb": "Standard_E32_v3",
+            "CPU-432Gb":"Standard E64_v3",
+            "GPU-56GB": "Standard_NC6_Promo",
+            "GPU-112GB":"Standard_NV12",
+            "GPU-224GB":"Standard_NV24"
+
+        },
+        "vmName": "[concat(parameters('vmName'),'-', parameters('cpu-gpu'))]",
+        "dataDisks": {
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: this is never used

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+            "lun": 0,
+            "createOption": "attach",
+            "caching": "None",
+            "writeAcceleratorEnabled": false,
+            "id": null,
+            "name": "[concat('Datadisk', parameters('vmName'))]",
+            "storageAccountType": null,
+            "diskSizeGB": null    
+        },
+        "dataDiskResources": {
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: this is never used

Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: it should be my mistake.. just PR a new version which creates the date disk and attached it as a lun

File: 201-vm-linux-jupyterhub/azuredeploy.json (hunk 2 above)
Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: My mistake i didnt attach the data disk this has been fixed

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 4]
```diff
@@ -287,14 +286,29 @@
             },
             "kind": "[variables('storageAccountKind')]"
         },
+        {
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: remove this and use implicit disks

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 5]
```diff
@@ -32,7 +32,7 @@
             "type": "string",
             "defaultValue": "CPU-112GB",
             "allowedValues": [
-                "CPU-8GB",
+               "CPU-8GB",
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: format the json, e.g. SHIFT+ALT+F in VS Code

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 6]
```diff
@@ -312,7 +326,24 @@
                         "offer": "UbuntuServer",
                         "sku": "18.04-LTS",
                         "version": "latest"
-                    }
+                    },
+                    "copy": [
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: remove the copy loop, you're only creating one disk

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 7]
```diff
@@ -325,6 +356,12 @@
                     "computerName": "[parameters('vmName')]",
                     "adminUsername": "[parameters('adminUsername')]",
                     "adminPassword": "[parameters('adminPassword')]"
+                },
+                "diagnosticsProfile": {
+                    "bootDiagnostics": {
+                        "enabled": true,
+                        "storageUri": "[concat('https://', variables('storageAccountName'), '.blob.core.windows.net/')]"
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: use reference() to get the uri

File: 201-vm-linux-jupyterhub/azuredeploy.json (hunk 5 above)
Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: resolved

File: 201-vm-linux-jupyterhub/azuredeploy.json (hunk 7 above)
Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: bringing alignment to 101-vm-linux-jupyerhub template but having larger machine sizes

File: 201-vm-linux-jupyterhub/azuredeploy.json (hunk 6 above)
Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: bringing alignment to 101-vm-linux-jupyerhub template but having larger machine sizes

File: 201-vm-linux-jupyterhub/azuredeploy.json (hunk 4 above)
Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: bringing alignment to 101-vm-linux-jupyerhub template but having larger machine sizes

File: 201-vm-linux-jupyterhub/azuredeploy.json [hunk 8]
```diff
@@ -359,8 +314,8 @@
                 },
                 "diagnosticsProfile": {
                     "bootDiagnostics": {
-                        "enabled": true,
-                        "storageUri": "[concat('https://', variables('storageAccountName'), '.blob.core.windows.net/')]"
+                        "enabled":  true ,
+                        "storageUri": "[concat(reference(concat('Microsoft.Storage/storageAccounts/', variables('storageAccountName')), '2016-01-01').primaryEndpoints.blob)]"
```
Comment (bmoore-msft) [inline 201-vm-linux-jupyterhub/azuredeploy.json]: reference(variables('storageAccountName')).primaryEndpoints.blob

Comment (leestott) [reply 201-vm-linux-jupyterhub/azuredeploy.json]: @bmoore-msft thanks changes made

**Pull request merged**